cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java

```diff
@@ -79,7 +101,62 @@ public void onFinish(HttpServiceResponder responder) throws Exception {
 
         @Override
         public void onError(HttpServiceResponder responder, Throwable failureCause) {
-          responder.sendStatus(500);
+          responder.sendString(500, failureCause.getMessage(), Charsets.UTF_8);
+        }
+      };
+    }
+
+    @GET
+    @Path("/download")
+    public void download(HttpServiceRequest request, HttpServiceResponder responder) {
+      // Responds with a content producer
+      KeyValueTable table = getContext().getDataset(HANDLER_TABLE_NAME);
+      responder.send(200, new DownloadHttpContentProducer(table), "text/plain");
+    }
+
+    @POST
+    @Path("/uploadDownload")
+    public HttpContentConsumer uploadDownload(HttpServiceRequest request, HttpServiceResponder responder) {
+      // Consume request with content consumer, then response with content producer
+      return new HttpContentConsumer() {
+        @Override
+        public void onReceived(ByteBuffer chunk, Transactional transactional) throws Exception {
+          // no-op
+        }
+
+        @Override
+        public void onFinish(HttpServiceResponder responder) throws Exception {
+          KeyValueTable table = getContext().getDataset(HANDLER_TABLE_NAME);
+          responder.send(200, new DownloadHttpContentProducer(table), "text/plain");
+        }
+
+        @Override
+        public void onError(HttpServiceResponder responder, Throwable failureCause) {
+          responder.sendString(500, failureCause.getMessage(), Charsets.UTF_8);
+        }
+      };
+    }
+
+    @PUT
+    @Path("/invalid")
+    public HttpContentConsumer invalid(HttpServiceRequest request, HttpServiceResponder responder) {
+      // Create a closure on the handler responder
+      final HttpServiceResponder finalResponder = responder;
+      return new HttpContentConsumer() {
+        @Override
+        public void onReceived(ByteBuffer chunk, Transactional transactional) throws Exception {
+          // no-op
+        }
+
+        @Override
+        public void onFinish(HttpServiceResponder responder) throws Exception {
+          // Use the closure responder instead of the given responder, it should fail
+          finalResponder.sendStatus(200);
```

Comment:
So this will throw an exception, which gets caught and the onError() is invoked. What if onError also uses the original responder? That will throw another exception, which will be logged and the connection gets closed without response?